oxo.py

```diff
@@ -1,26 +1,195 @@
+        """ The initialiser. Initialise any fields you need here.
+
+        :param grid_width:     width of board
+        :param grid_height:    height of board
+        :param win_row_amount:
+        """
+        self.grid_size = {"width": grid_width, "height": grid_height}
+        self.total_cells = grid_width * grid_height
+        self.grid = {}
+        self.win_row_amount = win_row_amount
 
     def get_square(self, x, y):
         """ Return 0, 1 or 2 depending on the contents of the specified square. """
-        raise NotImplementedError("TODO: implement get_square")
+        if (x, y) in self.grid:
+            return self.grid[(x, y)]
+        else:
+            return 0
 
     def set_square(self, x, y, mark):
         """ If the specified square is currently empty (0), fill it with mark and return True.
             If the square is not empty, leave it as-is and return False. """
-        raise NotImplementedError("TODO: implement set_square")
+        if (x, y) not in self.grid:
+            self.grid[(x, y)] = mark
+            return True
+        else:
+            return False
 
     def is_board_full(self):
         """ If there are still empty squares on the board, return False.
             If there are no empty squares, return True. """
-        raise NotImplementedError("TODO: implement is_board_full")
+        return len(self.grid) == self.total_cells
 
     def get_winner(self):
         """ If a player has three in a row, return 1 or 2 depending on which player.
             Otherwise, return 0. """
-        raise NotImplementedError("TODO: implement get_winner")
+        for player_index in range(1, 3, 1):
+            # Horizontal
+            winner = self.check_hoz(player_index)
+            if winner != 0:
+                return winner
+            # Vertical
+            winner = self.check_vert(player_index)
+            if winner != 0:
+                return winner
+            # Diagonals
+            winner = self.check_diagonals(player_index)
+            if winner != 0:
+                return winner
+        return 0
+
+    def check_hoz(self, player):
+        """Check the horizontally for a winner
+
+        :param player:  player index
+        :return:        winner id (0 if no winner)
+        """
+        # there is no point checking the board if the grid is not wide enough for the win condition
+        if self.win_row_amount > self.grid_size["width"]:
+            return 0
+
+        count_in_row = 0
+
+        for y in range(self.grid_size["height"]):
+            for x in range(self.grid_size["width"]):
+
+                # break if its not possible to get the required amount it a row
+                if count_in_row == 0 and self.grid_size["width"] - x < self.win_row_amount:
+                    break
+
+                # reset and continue if we have no input
+                if (x, y) not in self.grid:
+                    count_in_row = 0
+                    continue
```

Comment:
This all looks ok, however be wary of using `break` and `continue` too much as they can make code hard to follow. This could have been rewritten just with nested `if` statements for example.